File: backend/api/l1.py
from fastapi import APIRouter, HTTPException
from fastapi.responses import StreamingResponse
from pydantic import BaseModel
from typing import Optional
import json
from datetime import datetime

from backend.services.project_manager import get_project_manager
from backend.services.meeting_logger import MeetingLogger
from backend.core.registry import get_module
from backend.core.config import get_config

router = APIRouter()


class L1Input(BaseModel):
    idea: str = ""
    target_readers: str = ""
    core_appeal: str = ""
    style: str = ""
    outline: str = ""  # 兼容旧字段
    rough_outline: str = ""  # 新字段名，与 VisionDocument 一致
    world_setting: str = ""
    protagonist: str = ""
    golden_finger: str = ""
    hot_elements: str = ""
    expected_length: str = ""
    
    def get_outline(self) -> str:
        """获取大纲内容，优先使用 rough_outline"""
        return self.rough_outline or self.outline


@router.post("/projects/{project_id}/l1/generate")
async def generate_vision(project_id: str, data: L1Input):
    pm = get_project_manager()
    project = pm.get_project(project_id)
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    
    config = get_config()
    generator_cls = get_module("l1", "guided_form")
    generator = generator_cls()
    
    # 构建输入数据，统一字段名
    input_data = data.model_dump()
    input_data["outline"] = data.get_outline()  # 使用统一的大纲字段
    
    vision = generator.generate(input_data)
    
    vision_path = project.project_path / "outputs" / "l1_vision.json"
    vision_path.parent.mkdir(parents=True, exist_ok=True)
    
    vision_data = vision.model_dump()
    vision_data["created_at"] = datetime.now().isoformat()
    
    with open(vision_path, "w", encoding="utf-8") as f:
        json.dump(vision_data, f, ensure_ascii=False, indent=2)
    
    logger = MeetingLogger(project.project_path / "logs" / "meeting.db")
    logger.log_version(project_id, "l1", vision_data, "L1 vision generated")
    
    return {"status": "generated", "vision": vision_data}


@router.get("/projects/{project_id}/l1/vision")
async def get_vision(project_id: str):
    pm = get_project_manager()
    project = pm.get_project(project_id)
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    
    vision_path = project.project_path / "outputs" / "l1_vision.json"
    if not vision_path.exists():
        raise HTTPException(status_code=404, detail="Vision not found")
    
    with open(vision_path, "r", encoding="utf-8") as f:
        vision_data = json.load(f)
    
    return {"vision": vision_data}


@router.put("/projects/{project_id}/l1/vision")
async def update_vision(project_id: str, data: dict):
    pm = get_project_manager()
    project = pm.get_project(project_id)
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    
    vision_path = project.project_path / "outputs" / "l1_vision.json"
    if not vision_path.exists():
        raise HTTPException(status_code=404, detail="Vision not found")
    
    with open(vision_path, "r", encoding="utf-8") as f:
        vision_data = json.load(f)
    
    vision_data.update(data)
    vision_data["updated_at"] = datetime.now().isoformat()
    
    with open(vision_path, "w", encoding="utf-8") as f:
        json.dump(vision_data, f, ensure_ascii=False, indent=2)
    
    return {"status": "updated", "vision": vision_data}


# L1 聊天引导
class ChatMessage(BaseModel):
    role: str
    content: str


class L1ChatRequest(BaseModel):
    messages: list[ChatMessage]
    currentForm: dict = {}


class L1ChatResponse(BaseModel):
    reply: str
    extracted: dict = {}


@router.post("/projects/{project_id}/l1/chat")
async def l1_chat(project_id: str, req: L1ChatRequest):
    pm = get_project_manager()
    project = pm.get_project(project_id)
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    
    config = get_config()
    llm_cls = get_module("llm", config.llm.primary)
    llm = llm_cls()
    
    # 构建对话上下文
    system_prompt = """你是L1种子层的创作引导助手。你的任务是通过对话帮助用户梳理小说创意。

对话策略：
1. 每次只问一个问题，逐步引导
2. 根据用户回答，自然地追问下一个要素
3. 当收集到足够信息时，可以主动总结
4. 保持友好、专业的语气

需要收集的要素（按优先级）：
- 类型/题材（重生/玄幻/科幻等）
- 主角人设
- 金手指/核心能力
- 核心爽点
- 世界观背景
- 风格基调
- 粗略大纲

当前对话阶段：根据已有信息判断下一步该问什么。"""

    messages = [{"role": "system", "content": system_prompt}]
    
    for msg in req.messages:
        messages.append({"role": "user" if msg.role == "user" else "assistant", "content": msg.content})
    
    # 简单的信息提取逻辑
    extracted = {}
    last_user_msg = req.messages[-1].content if req.messages and req.messages[-1].role == "user" else ""
    
    # 提取类型
    genres = ["重生", "穿越", "玄幻", "修仙", "科幻", "都市", "历史", "架空", "网游", "末日"]
    for g in genres:
        if g in last_user_msg:
            extracted["genre"] = g
            break
    
    # 提取金手指关键词
    if "系统" in last_user_msg:
        extracted["golden_finger"] = "系统"
    elif "重生" in last_user_msg and "记忆" in last_user_msg:
        extracted["golden_finger"] = "前世记忆"
    
    # 提取核心梗（限制长度）
    if len(last_user_msg) > 10 and len(last_user_msg) < 100:
        extracted["idea"] = last_user_msg
    
    # 生成SSE流
    def generate():
        full_response = ""
        try:
            for chunk in llm.stream("", messages=messages, temperature=0.8):
                full_response += chunk
                yield f"data: {json.dumps({'type': 'chunk', 'content': chunk}, ensure_ascii=False)}\n\n"
            
            # 发送完成事件
            yield f"data: {json.dumps({'type': 'done', 'content': full_response, 'extracted': extracted}, ensure_ascii=False)}\n\n"
        except Exception as e:
            yield f"data: {json.dumps({'type': 'error', 'content': str(e)}, ensure_ascii=False)}\n\n"
    
    return StreamingResponse(generate(), media_type="text/event-stream")


@router.get("/projects/{project_id}/l1/draft")
async def get_draft(project_id: str):
    pm = get_project_manager()
    project = pm.get_project(project_id)
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    
    draft_path = project.project_path / "outputs" / "l1_draft.json"
    if not draft_path.exists():
        return {"draft": None}
    
    with open(draft_path, "r", encoding="utf-8") as f:
        draft_data = json.load(f)
    
    return {"draft": draft_data}


@router.post("/projects/{project_id}/l1/draft")
async def save_draft(project_id: str, data: dict):
    pm = get_project_manager()
    project = pm.get_project(project_id)
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    
    draft_path = project.project_path / "outputs" / "l1_draft.json"
    draft_path.parent.mkdir(parents=True, exist_ok=True)
    
    data["updated_at"] = datetime.now().isoformat()
    
    with open(draft_path, "w", encoding="utf-8") as f:
        json.dump(data, f, ensure_ascii=False, indent=2)
    
    return {"status": "saved"}


@router.get("/projects/{project_id}/l1/chat-history")
async def get_chat_history(project_id: str):
    pm = get_project_manager()
    project = pm.get_project(project_id)
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    
    history_path = project.project_path / "outputs" / "l1_chat_history.json"
    if not history_path.exists():
        return {"messages": []}
    
    with open(history_path, "r", encoding="utf-8") as f:
        history_data = json.load(f)
    
    return {"messages": history_data.get("messages", [])}


@router.post("/projects/{project_id}/l1/chat-history")
async def save_chat_history(project_id: str, data: dict):
    pm = get_project_manager()
    project = pm.get_project(project_id)
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    
    history_path = project.project_path / "outputs" / "l1_chat_history.json"
    history_path.parent.mkdir(parents=True, exist_ok=True)
    
    history_data = {
        "messages": data.get("messages", []),
        "updated_at": datetime.now().isoformat()
    }
    
    with open(history_path, "w", encoding="utf-8") as f:
        json.dump(history_data, f, ensure_ascii=False, indent=2)
    
    return {"status": "saved"}

File: backend-node/src/routes/l1.ts
// L1 种子层路由
import { Express, Request, Response } from 'express';
import * as fs from 'fs';
import * as path from 'path';
import { getProjectManager } from '../services/project-manager';
import { getLLM } from '../services/llm';
import { VisionDocument } from '../protocols';
import { SSEWriter } from '../utils/sse';

const L1_PROMPT = `你是一位创作顾问，帮助用户梳理创意愿景，产出清晰的故事种子文档。

请根据用户提供的以下信息，生成一份结构化的《故事愿景文档》：

用户输入：
{user_input}

请按以下格式输出：

# 故事愿景文档

## 核心梗
[一句话概括故事核心创意]

## 阅读契约
- 目标读者：[男频/女频/年龄层]
- 核心爽点：[装逼/升级/恋爱/悬疑等]
- 风格基调：[轻松/暗黑/热血/治愈等]

## 粗略大纲
1. 开篇：[...]
2. 发展：[...]
3. 高潮：[...]
4. 结局：[...]

## 核心设定
- 世界观：[...]
- 主角人设：[...]
- 金手指/核心道具：[...]

## 热点/潮流元素（如有）
- 热点：[...]
- 融合方式：[...]
- 时效性评估：[...]

## 预期字数
[长篇/中篇/短篇，预估字数]

注意：
- 不评判创意好坏，只负责结构化
- 如果用户信息不足，根据上下文合理补充
- 输出要简洁，不超过500字`;

export function registerL1Routes(app: Express): void {
  const pm = getProjectManager();

  // 获取默认提示词
  app.get('/api/projects/:projectId/l1/prompt', (req: Request, res: Response) => {
    const formParts: string[] = [];
    // 返回模板，用 {field} 占位
    const promptTemplate = `你是一位创作顾问，帮助用户梳理创意愿景，产出清晰的故事种子文档。

用户提供的信息：
{user_input}

请根据以上信息，生成一份完整的《故事愿景文档》。要求：
1. 直接输出文档内容，不要有多余的解释
2. 使用 markdown 格式
3. 包含以下部分：核心梗、阅读契约、粗略大纲、核心设定、热点元素、预期字数
4. 内容要具体、有指导意义，不要泛泛而谈
5. 如果用户信息不足，合理补充但要符合用户意图`;

    res.json({ prompt: promptTemplate });
  });

  // L1 聊天（SSE 流式）
  app.post('/api/projects/:projectId/l1/chat', async (req: Request, res: Response) => {
    const project = pm.getProject(req.params.projectId);
    if (!project) {
      return res.status(404).json({ error: 'Project not found' });
    }

    const { messages = [], currentForm = {} } = req.body;
    const writer = new SSEWriter(res);

    try {
      const llm = getLLM();
      
      // 读取愿景文档（如果存在）
      let visionContent = '';
      const visionPath = path.join(project.path, 'outputs', 'l1_vision.json');
      if (fs.existsSync(visionPath)) {
        try {
          const visionData = JSON.parse(fs.readFileSync(visionPath, 'utf-8'));
          visionContent = visionData.content || JSON.stringify(visionData, null, 2);
        } catch {}
      }
      
      // 构建聊天历史
      const chatHistory = messages.map((m: Record<string, string>) => 
        `${m.role === 'user' ? '用户' : 'AI'}: ${m.content}`
      ).join('\n');

      let prompt = `你是创作顾问，帮助用户梳理创意。

聊天历史：
${chatHistory}`;

      // 如果有愿景文档，添加到提示词
      if (visionContent) {
        prompt += `

=== 当前愿景文档 ===
${visionContent}

注意：用户已经生成了愿景文档，请根据文档内容回答问题。如果用户询问愿景文档内容，请基于上述文档回答。`;
      }

      prompt += `

请继续引导用户完善创意。保持友好、专业的语气，每次只问一个问题，逐步引导。`;

      let fullContent = '';
      let thinking = '';

      for await (const chunk of llm.stream(prompt, { thinking: true, thinkingBudget: 5000 })) {
        if (chunk.type === 'thinking') {
          thinking += chunk.content;
          writer.writeData({ type: 'thinking', content: chunk.content });
        } else {
          fullContent += chunk.content;
          writer.writeData({ type: 'chunk', content: chunk.content });
        }
      }

      writer.writeData({ 
        type: 'done', 
        content: fullContent,
        thinking
      });
    } catch (error) {
      writer.write('error', { message: String(error) });
    } finally {
      writer.close();
    }
  });

  // 生成愿景文档（SSE 流式）
  app.post('/api/projects/:projectId/l1/generate', async (req: Request, res: Response) => {
    const project = pm.getProject(req.params.projectId);
    if (!project) {
      return res.status(404).json({ error: 'Project not found' });
    }

    const writer = new SSEWriter(res);

    try {
      const llm = getLLM();
      
      // 把表单内容合并成一个提示字符串
      const formParts: string[] = [];
      if (req.body.idea) formParts.push(`创意/脑洞：${req.body.idea}`);
      if (req.body.target_readers) formParts.push(`目标读者：${req.body.target_readers}`);
      if (req.body.core_appeal) formParts.push(`核心爽点：${req.body.core_appeal}`);
      if (req.body.style) formParts.push(`风格基调：${req.body.style}`);
      if (req.body.rough_outline) formParts.push(`粗略大纲：${req.body.rough_outline}`);
      if (req.body.world_setting) formParts.push(`世界观设定：${req.body.world_setting}`);
      if (req.body.protagonist) formParts.push(`主角人设：${req.body.protagonist}`);
      if (req.body.golden_finger) formParts.push(`金手指/核心道具：${req.body.golden_finger}`);
      if (req.body.hot_elements) formParts.push(`热点元素：${req.body.hot_elements}`);
      if (req.body.expected_length) formParts.push(`预期字数：${req.body.expected_length}`);
      
      const userInput = formParts.length > 0 
        ? formParts.join('\n') 
        : '用户未提供具体信息，请根据创意方向生成';

      // 使用用户自定义提示词或默认提示词
      let prompt = req.body.custom_prompt || `你是一位创作顾问，帮助用户梳理创意愿景，产出清晰的故事种子文档。

用户提供的信息：
{user_input}

请根据以上信息，生成一份完整的《故事愿景文档》。要求：
1. 直接输出文档内容，不要有多余的解释
2. 使用 markdown 格式
3. 包含以下部分：核心梗、阅读契约、粗略大纲、核心设定、热点元素、预期字数
4. 内容要具体、有指导意义，不要泛泛而谈
5. 如果用户信息不足，合理补充但要符合用户意图`;

      // 替换 {user_input} 占位符
      prompt = prompt.replace('{user_input}', userInput);

      // 流式输出
      let fullContent = '';
      let thinkingContent = '';

      for await (const chunk of llm.stream(prompt, { 
        thinking: true, 
        thinkingBudget: 10000 
      })) {
        if (chunk.type === 'thinking') {
          thinkingContent += chunk.content;
          writer.writeData({ type: 'thinking', content: chunk.content });
        } else {
          fullContent += chunk.content;
          writer.writeData({ type: 'content', content: chunk.content });
        }
      }

      // 保存原始 markdown
      const visionPath = path.join(project.path, 'outputs', 'l1_vision.json');
      fs.mkdirSync(path.dirname(visionPath), { recursive: true });
      fs.writeFileSync(visionPath, JSON.stringify({
        content: fullContent,
        thinking: thinkingContent,
        createdAt: new Date().toISOString()
      }, null, 2), 'utf-8');

      writer.writeData({ type: 'done', content: fullContent, thinking: thinkingContent });
    } catch (error) {
      writer.write('error', { message: String(error) });
    } finally {
      writer.close();
    }
  });

  // 获取愿景文档
  app.get('/api/projects/:projectId/l1/vision', (req: Request, res: Response) => {
    const project = pm.getProject(req.params.projectId);
    if (!project) {
      return res.status(404).json({ error: 'Project not found' });
    }

    const visionPath = path.join(project.path, 'outputs', 'l1_vision.json');
    if (!fs.existsSync(visionPath)) {
      return res.status(404).json({ error: 'Vision not found' });
    }

    const data = JSON.parse(fs.readFileSync(visionPath, 'utf-8'));
    // 兼容新旧格式
    const vision = data.content ? data : { content: JSON.stringify(data, null, 2) };
    res.json({ vision });
  });

  // 创建/保存愿景文档
  app.post('/api/projects/:projectId/l1/vision', (req: Request, res: Response) => {
    const project = pm.getProject(req.params.projectId);
    if (!project) {
      return res.status(404).json({ error: 'Project not found' });
    }

    const visionPath = path.join(project.path, 'outputs', 'l1_vision.json');
    fs.mkdirSync(path.dirname(visionPath), { recursive: true });
    fs.writeFileSync(visionPath, JSON.stringify({
      content: req.body.content || req.body,
      updatedAt: new Date().toISOString()
    }, null, 2), 'utf-8');

    res.json({ status: 'saved' });
  });

  // 更新愿景文档
  app.put('/api/projects/:projectId/l1/vision', (req: Request, res: Response) => {
    const project = pm.getProject(req.params.projectId);
    if (!project) {
      return res.status(404).json({ error: 'Project not found' });
    }

    const visionPath = path.join(project.path, 'outputs', 'l1_vision.json');
    fs.mkdirSync(path.dirname(visionPath), { recursive: true });
    fs.writeFileSync(visionPath, JSON.stringify({
      ...req.body,
      updatedAt: new Date().toISOString()
    }, null, 2), 'utf-8');

    res.json({ status: 'updated' });
  });

  // 保存草稿
  app.get('/api/projects/:projectId/l1/draft', (req: Request, res: Response) => {
    const project = pm.getProject(req.params.projectId);
    if (!project) {
      return res.status(404).json({ error: 'Project not found' });
    }

    const draftPath = path.join(project.path, 'outputs', 'l1_draft.json');
    if (fs.existsSync(draftPath)) {
      const draft = JSON.parse(fs.readFileSync(draftPath, 'utf-8'));
      res.json({ draft });
    } else {
      res.json({ draft: {} });
    }
  });

  app.post('/api/projects/:projectId/l1/draft', (req: Request, res: Response) => {
    const project = pm.getProject(req.params.projectId);
    if (!project) {
      return res.status(404).json({ error: 'Project not found' });
    }

    const draftPath = path.join(project.path, 'outputs', 'l1_draft.json');
    fs.mkdirSync(path.dirname(draftPath), { recursive: true });
    fs.writeFileSync(draftPath, JSON.stringify(req.body, null, 2), 'utf-8');

    res.json({ status: 'saved' });
  });

  // 聊天历史
  app.get('/api/projects/:projectId/l1/chat-history', (req: Request, res: Response) => {
    const project = pm.getProject(req.params.projectId);
    if (!project) {
      return res.status(404).json({ error: 'Project not found' });
    }

    const historyPath = path.join(project.path, 'outputs', 'l1_chat_history.json');
    if (fs.existsSync(historyPath)) {
      const history = JSON.parse(fs.readFileSync(historyPath, 'utf-8'));
      res.json({ history });
    } else {
      res.json({ history: [] });
    }
  });

  app.post('/api/projects/:projectId/l1/chat-history', (req: Request, res: Response) => {
    const project = pm.getProject(req.params.projectId);
    if (!project) {
      return res.status(404).json({ error: 'Project not found' });
    }

    const historyPath = path.join(project.path, 'outputs', 'l1_chat_history.json');
    fs.mkdirSync(path.dirname(historyPath), { recursive: true });
    fs.writeFileSync(historyPath, JSON.stringify(req.body, null, 2), 'utf-8');

    res.json({ status: 'saved' });
  });
}

function parseVisionResponse(response: string, originalInput: Record<string, string>): VisionDocument {
  const extractSection = (name: string): string => {
    const regex = new RegExp(`##\\s*${name}[^\\n]*\\n([\\s\\S]*?)(?=##|$)`, 'i');
    const match = response.match(regex);
    return match ? match[1].trim() : '';
  };

  const extractField = (text: string, field: string): string => {
    const lines = text.split('\n');
    for (const line of lines) {
      if (line.includes(field)) {
        const parts = line.split(/[：:]/);
        return parts.length > 1 ? parts[1].trim() : '';
      }
    }
    return '';
  };

  const contract = extractSection('阅读契约');
  const setting = extractSection('核心设定');

  // 返回下划线格式，与前端兼容
  return {
    core_idea: extractSection('核心梗') || originalInput.idea || originalInput.core_idea || '',
    target_readers: extractField(contract, '目标读者') || originalInput.targetReaders || originalInput.target_readers || '',
    core_appeal: extractField(contract, '核心爽点') || originalInput.coreAppeal || originalInput.core_appeal || '',
    style: extractField(contract, '风格基调') || originalInput.style || '',
    rough_outline: extractSection('粗略大纲') || originalInput.roughOutline || originalInput.rough_outline || originalInput.outline || '',
    world_setting: extractField(setting, '世界观') || originalInput.worldSetting || originalInput.world_setting || '',
    protagonist: extractField(setting, '主角人设') || originalInput.protagonist || '',
    golden_finger: extractField(setting, '金手指') || extractField(setting, '核心道具') || originalInput.goldenFinger || originalInput.golden_finger || '',
    hot_elements: extractSection('热点/潮流元素') || originalInput.hotElements || originalInput.hot_elements || '',
    expected_length: extractSection('预期字数') || originalInput.expectedLength || originalInput.expected_length || ''
  } as VisionDocument;
}
